lib/karto_sdk/include/karto_sdk/Karto.h

```diff
@@ -6285,9 +6362,17 @@ namespace karto
           kt_int32u* pCellHitCntPtr = m_pCellHitsCnt->GetDataPointer();
 
           // increment cell pass through and hit count
-          pCellPassCntPtr[index]++;
-          pCellHitCntPtr[index]++;
-
+          //TODO:make the maxCntLimit a variable modded by .yaml
+          if (m_dynamicEnvMode->GetValue()){
```

Comment:
Re:saturation:
That is the intention, currently if you are in a space for a long period of time and the environment is relatively static, it will take also a long period of time for the map to reflect any change in the environment. For example if the robot is in a closed room and after a long time someone opens a door, the map will get updated with what is past the door, but the door itself will remain there until a proportional amount (relative to the time the door was closed) to remove the door from the map. The saturation aims to reduce the time for the door to update to a maximum time based on the saturation value.

Re:multiplier
This multiplier was added because (with the lidard that i am testing) there were too many false negatives, removing features of the map that were 'noisy' due to small features. I agree that this approach is not ideal in a probabilistic approach, but is very useful when tuning for dynamic environments.

Reply:
Well for saturation, you can get into a spot where a cell has reached both maximum counts for hit and pass so its static no matter what data comes in the future. Since there's no removal you're going to basically just remain the same for all time

The multiplier cannot stay, that's a determent to mapping. "false negatives" I'm not sure I know what you're saying but it sounds like this was a patch for your the saturation issue

Reply:
I will elaborate on the multiplier with some examples:
This is how the map looks when the multiplier is set to `1` 
![image](https://user-images.githubusercontent.com/4633760/68652317-1e61b180-0521-11ea-9d5b-636f91279e1d.png)

This is how the map looks when the multiplier is set to `2`
![image](https://user-images.githubusercontent.com/4633760/68652414-4d782300-0521-11ea-8b83-8cec15562b0f.png)

For both cases, the red dots indicate the laser scans.

I think that the reason why the occupied cells in the `1` case seem not to be there is because measurement error from the lidar. As when a cell does not record a hit, the hit count is reduced, the cells are kept mostly with a low number of hits. I think that maybe a better lidar with more measurement precision wont experience this issue.

Because of this I added the option to add a multiplier to the hit count, so that when a cell is hit, it has more impact than when a cell is missed.

Also the way that I understand the TraceLine function, the cells traced (that will score passes, an in my modified case, remove hits) can be shared between two contiguous end points of a lidar sample. This will increase the number of times a cell hit count is reduced, therefore the multiplier to add more hits every time an end point is valid.

Reply:
So your example is definitely problematic. You cannot have a single laser scan have such high likelihood that you fully mark the map after a single scan. That defeats the probablistic approaches to make sure over multiple measurements across a scene they're consistent to eventually mark. That will treat you **very** poorly in any medium-sized environment or if you have wheel slippage. This would only work if you had exactly prefer positioning, which is the problem that SLAM is trying to solve, you'll end up degrading both the map quality and the positioning substantially. 

It is a feature that it takes multiple measurements to fully update a cell. It should never be the case that a one-shot laser scan should mark the visible environment so fully. 

I'm still just not sure what you're getting at. You have a counter for both the hits and the misses, why do you need to "remove" hits? You can change the ratio, or have a different computed metric in "dynamic map mode" that isn't misses/hits (maybe its hits/(hits+misses) or something that allows faster updates with dynamic changes). I don't see a situation when "removing" actual measurements that read things is going to be generally good for mapping. I think you just want to combine the set of information differently and set a new ratio value.

Hit / pass makes it so pass needs to grow to break some threshold. If you set this ratio lower, it will update quicker, that parameter already exists just needs to be exposed as we talked about in the ticket. I'm still not convinced that that won't fix your concerns. If hit / pass isn't the right metric, hit / (sum(hit, pass)) may be. 

Either way, "dynamic environments" cannot just blindly take measurements and place them in the occupancy grid, that's not mapping that's just obstacle avoidance. There needs to be a few update cycles to let us know that those measurements are good and that object is where we think it is before inserting. If the problem is you want to insert obstacles faster, I don't see where lowering this ratio doesn't make it faster. It won't be immediate, but it should never be immediate.

You should never be working in environments with doors or where things move alot and using the SLAM generated map as your only input for collision avoidance. It seems like you're having this issue because you have the map as an input as the planning space instead of using the current state of your surroundings for the planning space and the map generated by SLAM as a weight into it for longer distances (ei apply a weight to NOT go into a wall/door, but allow it to do so. Lowest cost path won't trace back through walls but if you need to go through door it'll do it since there's no other valid path).

Reply:
RE:`You cannot have a single laser scan have such high likelihood that you fully mark the map after a single scan` && `It is a feature that it takes multiple measurements to fully update a cell. It should never be the case that a one-shot laser scan should mark the visible environment so fully.`
I dont know to what are you referring here, the cell only gets updated when it has had at least `2` passes (defined by `m_pMinPassThrough `). I have not made any modification (that I am aware of) that alters how many passes a cell needs to have before being able to decide if it is occupied or not. Could you elaborate please?

RE:`You have a counter for both the hits and the misses, why do you need to "remove" hits? `
I did not add any misses counters, the hit and pass counters already existed the only mod I did to it was to be able to remove the number of hits if a ray passed over that cell and it did not "hit" it.

I found interesting that the pass counter is updated twice in the RayTrace function, once when it is at the end of a valid end point of a laser beam, and another time when it is within the trayectory of the TraceLine line function. The TraceLine function is the one that I am modding to remove hits when the laser beam passed through a cell and was not the end point of the laser beam. 

RE:`Hit / pass makes it so pass needs to grow to break some threshold. ` 
Agree, the problem is that the original configuration made the threshold grow indefinite if the cell had a high number of passes or passes and hits, which will eventually happen if the robot is operating on an environment for long periods of time. That is the main reason why I added a limit to the amount of passes and hits a cell can have, so there is an upper limit to the amount of time it will take to update a cell.

RE:` You can change the ratio, or have a different computed metric in "dynamic map mode" that isn't misses/hits (maybe its hits/(hits+misses) or something that allows faster updates with dynamic changes).`
This is an interesting idea, I am not counting the misses, but I can think of a way to make a mod in this direction.

RE:`Hit / pass makes it so pass needs to grow to break some threshold.`
The problem is still the limit on how many hit/passes a cell can have

RE:` If you set this ratio lower, it will update quicker, that parameter already exists just needs to be exposed as we talked about in the ticket. `
Both variables, plus the ones I added to make my mod are exposed so that they can be modified via the .yaml config file.

RE:`Either way, "dynamic environments" cannot just blindly take measurements and place them in the occupancy grid, that's not mapping that's just obstacle avoidance.`
I understand the difference, I am using move base and the local cost map for that. The global cost map depends on the map topic so it can make the global path planning, while the local costmap uses the scan topic to avoid nearby obstacles that are not in the map (or are not there yet)

RE:`Either way, "dynamic environments" cannot just blindly take measurements and place them in the occupancy grid, that's not mapping that's just obstacle avoidance.`
I am not aware of any modification that I have made that inserts measurement into the map. Can you elaborate on this?

RE:`if the problem is you want to insert obstacles faster, I don't see where lowering this ratio doesn't make it faster. It won't be immediate, but it should never be immediate.`
Again, the time it takes to update is proportional to the count of passes and hits of the cell. If the robot scans that cell often for a long period of time, it will take a proportional amount of time to update once the cell has changed its occupancy in the real world map. 

RE:`You should never be working in environments with doors or where things move alot and using the SLAM generated map as your only input for collision avoidance`
As stated before I am not using the slam map for collision avoidance, that is done by the local cost map of move_base. The idea is for the map to react faster to changes in the map, example opening and closing a door, a new object tin the environment, like a box placed in the middle of the room, etc, not moving obstacles like people walking by or other robots.

RE:`It seems like you're having this issue because you have the map as an input as the planning space instead of using the current state of your surroundings for the planning space and the map generated by SLAM as a weight into it for longer distances (ei apply a weight to NOT go into a wall/door, but allow it to do so. Lowest cost path won't trace back through walls but if you need to go through door it'll do it since there's no other valid path).`
To be able to do that kind of behaviour I would need to do a door identifier on the map generated by the slam, and pray that the door is not closed so it looks like part of the wall I suppose.

EDIT: I have exposed all of the variables added, plus the threshold and the min number of passes so they can be ser in the config file. Would you like me to send you a bag file with my setup for you to run it and see how it works?